### PR TITLE
Pack optimizations

### PR DESCRIPTION
**Pull request by klaussfreire**
Tons of optimizations, some format changes too

### CODE REVIEW

File: sharedbuffers/mapped_struct.py [hunk 1]
```diff
@@ -421,12 +484,19 @@ def unpack_from(cls, buf, offs, idmap = None):
...
+            fs_type = pbuf[offs]
+            if fs_type == 'm' or fs_type == 'M':
                 # inline bitmap
-                if cython.compiled and offs+7 >= pybuf.len:
+                if fs_type == 'm':
+                    fs_size = 7
+                elif fs_type == 'M':
+                    fs_size = 15
+                else:
+                    raise ValueError("Unknown set type %r" % fs_type)
```
Comment (llogiudice) [inline sharedbuffers/mapped_struct.py:496]: isn't this check a bit unnecessary, since you've already tested for those types above ?

File: sharedbuffers/mapped_struct.py [hunk 2]
```diff
@@ -2667,18 +2904,29 @@ def init(self, slot_types = None, slot_keys = None, alignment = 8, pack_buffer_s
...
+    def _get_bitmaps(self, obj, get_values=False):
         has_bitmap = 0
         none_bitmap = 0
+        if get_values:
+            values = []
         for i,slot in enumerate(self.slot_keys):
-            if hasattr(obj, slot):
+            try:
+                val = getattr(obj, slot)
+            except AttributeError:
```
Comment (llogiudice) [inline sharedbuffers/mapped_struct.py:2939]: is this actually better than doing a ```hasattr``` test ? I'd have thought that exception handling was always slower.

File: sharedbuffers/mapped_struct.py [hunk 3]
```diff
@@ -2667,18 +2904,29 @@ def init(self, slot_types = None, slot_keys = None, alignment = 8, pack_buffer_s
...
+                if val is None:
                     none_bitmap |= cython.cast(cython.ulonglong, 1) << i
+                elif get_values:
+                    values.append(val)
         present_bitmap = has_bitmap & ~none_bitmap
-        return has_bitmap, none_bitmap, present_bitmap
+        if get_values:
+            return has_bitmap, none_bitmap, present_bitmap, values
+        else:
+            return has_bitmap, none_bitmap, present_bitmap
```
Comment (llogiudice) [inline sharedbuffers/mapped_struct.py:2951]: shouldn't this return a list in its fourth return value so the unpacking in ```get_packable``` doesn't fail ?

File: sharedbuffers/mapped_struct.py (hunk 1 above)
Comment (klaussfreire) [reply sharedbuffers/mapped_struct.py:496]: Yes, but it's free and it may save your life if the code that makes it unnecessary changes

File: sharedbuffers/mapped_struct.py (hunk 2 above)
Comment (klaussfreire) [reply sharedbuffers/mapped_struct.py:2939]: Exception handling in Cython is quite fast. That said, I haven't timed it on basic attributes. But it's most likely faster due to another subtle reason: hasattr on properties (ie: all proxied properties) has to invoke the getter and check if it rases AttributeError. That is, this exact code, only it discards the value if it succeeds. This is faster on properties because it avoids one whole invocation to the getter, which can be quite expensive.

File: sharedbuffers/mapped_struct.py (hunk 3 above)
Comment (klaussfreire) [reply sharedbuffers/mapped_struct.py:2951]: The whole point of `get_values` is to change the shape of the returned tuple

File: tests/mapped_struct.py [hunk 4]
```diff
@@ -336,8 +336,8 @@ def testUnpackInto(self):
             px = self.schema.pack(x)
             dx = self.schema.unpack(px, proxy_into = dx)
             for k,v in TEST_VALUES.iteritems():
-                self.assertTrue(hasattr(dx, k))
                 self.assertEqual(getattr(dx, k), v)
+                self.assertTrue(hasattr(dx, k))
```
Comment (tzulberti-jampp) [inline tests/mapped_struct.py]: Why did you change the order of this? It should check that it has the attribute before checking the value

File: sharedbuffers/mapped_struct.py [hunk 5]
```diff
@@ -441,72 +511,132 @@ def unpack_from(cls, buf, offs, idmap = None):
...
-                buf[offs] = 'q'
-                dtype = 'l'
-            if dtype == 'l':
-                buf[offs+1:offs+8] = struct.pack('<Q', objlen)[:7]
+        if isinstance(obj, npndarray):
+            all_int = all_intlong = all_float = 0
+            obj_dtype = obj.dtype
+            if obj_dtype.isbuiltin:
+                dtype = obj_dtype.char
+                if dtype in ('l','I','i','H','h','B','b'):
```
Comment (tzulberti-jampp) [inline sharedbuffers/mapped_struct.py]: nit: space between the tuple values?

File: sharedbuffers/mapped_struct.py [hunk 6]
```diff
@@ -532,29 +664,77 @@ def pack_into(cls, obj, buf, offs, idmap = None, implicit_offs = 0,
             else:
                 widmap = None
 
+            # Get a sense of whether we can use narrow pointers
+            min_offs = max_offs = 0
+            for i,x in enumerate(obj):
```
Comment (tzulberti-jampp) [inline sharedbuffers/mapped_struct.py]: nit: space between `i,x`

File: tests/mapped_struct.py (hunk 4 above)
Comment (klaussfreire) [reply tests/mapped_struct.py]: I was doing some tests and forgot to revert that

File: CHANGELOG.rst [hunk 7]
```diff
@@ -45,6 +45,11 @@ Changes
   perhaps considerably. Proper identity was implemented for proxied
   containers though.
 - Shrink some buffers by employing narrow pointers where possible
+- Optimized sequence packing
+- New wide bitmap frozenset format allows more frozensets to be packed
+  as bitmaps.
+- Offsets are Py_ssize_t now. That shouldn't be a noticeable change,
+  unless you've got more storage than they universe.
```
Comment (tzulberti-jampp) [inline CHANGELOG.rst]: Shouldn't it be `than the universe`?

File: sharedbuffers/mapped_struct.py [hunk 8]
```diff
@@ -421,12 +485,19 @@ def unpack_from(cls, buf, offs, idmap = None):
...
         try:
-            if pbuf[offs] == 'm':
+            fs_type = pbuf[offs]
+            if fs_type == 'm' or fs_type == 'M':
                 # inline bitmap
-                if cython.compiled and offs+7 >= pybuf.len:
+                if fs_type == 'm':
+                    fs_size = 7
+                elif fs_type == 'M':
+                    fs_size = 15
```
Comment (tzulberti-jampp) [inline sharedbuffers/mapped_struct.py:494]: Shouldn't this be 14? Because the for in line 500 is adding a +1.

Comment (klaussfreire) [reply sharedbuffers/mapped_struct.py:494]: No, `M` has 8 bytes more than `m`. The loop in 500 will go all the way up to i=14, which means `pbuf[offs+1+14] == pbuf[offs+15]` which is the last byte of a 16-byte bitmap.

**Comment by klaussfreire** [comment]
I'm going to stop adding stuff in here to incentivize reviews.

So... @tzulberti-jampp , @gefarion , @llogiudice , @npazos , it's your ball now

**Comment by klaussfreire** [comment]
Ping

File: sharedbuffers/mapped_struct.py [hunk 9]
```diff
@@ -385,26 +435,40 @@ def pack_into(cls, obj, buf, offs, idmap = None, implicit_offs = 0):
...
                         cbuf[offs+1+ix/8] |= 1 << (ix & 7)
                 else:
-                    buf[offs:offs+1] = 'm'
-                    buf[offs+1:offs+8] = '\x00\x00\x00\x00\x00\x00\x00'
+                    if maxval < 56:
+                        buf[offs:offs+1] = 'm'
+                        buf[offs+1:offs+8] = '\x00\x00\x00\x00\x00\x00\x00'
+                    else:
+                        buf[offs:offs+1] = 'M'
+                        buf[offs+1:offs+15] = '\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00'
```
Comment (llogiudice) [inline sharedbuffers/mapped_struct.py:454]: I think there's a mismatch here.
When ```maxval < 56```, you write 7 bytes, using offsets 1 to 8
When ```maxval >= 56```, you write 15 bytes, using offsets 1 to 15

I think in the second case, you have to use offsets 1 to 16 (And also update offs accordingly below).

**Review by tzulberti-jampp — APPROVED** [review]

**Review by llogiudice — APPROVED** [review]
